feat: code smell command, display commands error in chat

abeatrix opened the pull request:
RE feedback from beyang:

- Could we add tab-to-complete behavior, so if the user has typed in a partial match for the command name (e.g., “/expla”) and then hits Tab, we complete the command name (in this case “/explain”)?
- I noticed nothing happens if I execute /explain without anything selected. It seems intuitive to me to use the currently open file as context in this case.
- I think it makes sense to display any errors (like you need to have something selected) as a response in the chat, rather than as a separate error pop-up

This PR includes the following change:
1. for the default `/explain` & `/smell` command, run on visible content when there is no selection in the current text document
1. Display error message from commands in chat as an error message from the assistant when selection is required for a command (e.g. /doc and /test) but currently missing:
1. log basic telemetryService for commands
1. also cleaned up the custom-prompt recipe that all commands depend on.
1. show error for invalid commands:
<img width="515" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/85e86e59-916c-465e-9fd0-3ce4098e8dc3">
1. add `/smell`, command for code smell as the default command
1. display the file name and selection range when available
<img width="1363" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/0d643db7-0499-4c9f-bef7-dec963c8f5c2">

#### Some UI changes were moved to separate PRs

See moved to https://github.com/sourcegraph/cody/pull/606 & https://github.com/sourcegraph/cody/pull/648
- add tab-to-complete +  enter-to-complete behavior in the command list
- update the walkthrough to replace recipes with commands and add a new commands section:
- add short instructions on how to use commands with the new menu. The link also works as a button to open the command menu:
- update shortcut for commands menu to `alt+c` as it was using the same key as `cody.fixup.new`

## Test plan

<!-- Required. See https://docs.sourcegraph.com/dev/background-information/testing_principles. -->

See the complete demo:


https://github.com/sourcegraph/cody/assets/68532117/64f18b91-1fa4-4bfa-ae8a-acbce6360a39

1. start cody from this branch
2. run `/doc` without code selection - error expected
3. run non-exisit command:`/exit` - error expected
4. run `/explain` and `smell` without code selection - both show work with visible content
5. clicking on the `Cody Menu`button in the cody intro should open the command menu for you
7. shortcut `alt+c` should bring up the command menu
8. run commands from the right-click menu, everything should work accordingly

toolmantim commented on lib/ui/src/Chat.tsx:
```diff
@@ -271,6 +271,10 @@ export const Chat: React.FunctionComponent<ChatProps> = ({
                 }
+                if (event.key === 'Tab' && selectedChatCommand > -1) {
+                    const newInput = displayCommands?.[selectedChatCommand]?.[1]?.slashCommand
+                    setFormInput(newInput || formInput)
```
Needs a `event.preventDefault()` to prevent default behaviour (moving the focus away from the input)

toolmantim reviewed (commented):
I gave the tab complete behaviour a go locally, and looks like it needs a small fix to allow you to hit enter after hitting tab to complete.

> I noticed nothing happens if I execute /explain without anything selected. It seems intuitive to me to use the currently open file as context in this case.

I don't think we want to do this so generically — it looks like if you use `/document` with no code selected (but a file open) it now tries to explain the whole file?

And if we scope it to explain: how good is the experience explaining a whole file, vs nudging them to select a specific bit of code? Should we consider cursor position too?

> I think it makes sense to display any errors (like you need to have something selected) as a response in the chat, rather than as a separate error pop-up

Yeah, if you're triggering it from the chat window that is nice. Not sure if we want it as permanent part of the transcript though? More like a small error overlay that pops up above the compose box and hides after 5 seconds?

When you trigger stuff in-editor, without the chat window open, it would be nice if we didn't focus our chat pane if we didn't need to (add documentation, do an edit, etc). In that case, we'd want to keep the current behaviour of a native notification.

(Maybe worth breaking out the small tab completion fix from the rest?)

abeatrix commented:
> I don't think we want to do this so generically — it looks like if you use /document with no code selected (but a file open) it now tries to explain the whole file?

And if we scope it to explain: how good is the experience explaining a whole file, vs nudging them to select a specific bit of code? Should we consider cursor position too?

@toolmantim I think i used the wrong selection. Instead of selection or entire file, it should be selection or visible content, do you think that makes more sense? (Will take a look and make an update tomorrow if that sounds good to you)

abeatrix commented:
@toolmantim and what do you think about using visible content for `/doc` and `/test` if no selection was made?

abeatrix commented:
( removing beyang as reviewer for now )

toolmantim commented:
>> And if we scope it to explain: how good is the experience explaining a whole file, vs nudging them to select a specific bit of code? Should we consider cursor position too?

> I think i used the wrong selection. Instead of selection or entire file, it should be selection or visible content, do you think that makes more sense? (Will take a look and make an update tomorrow if that sounds good to you)

I'm not sure! I didn't meant to suggest the code in the PR was wrong… it definitely works as advertised.

I was more just wanting to check if the results it was returning was good? I was getting mixed results when I quickly tested it locally, and wasn't sure if it was a default path we wanted to push people down.

But I think what you've done matches what @beyang was suggesting, and it matches the context UI that's below the message box (the part that shows the current file)

Maybe it would be better if we took into account editor cursor as well (and make sure the context UI updates when you move lines)? That would hugely help for `/document`, where you have to select a whole line (otherwise it tries to document the whole file). And if you have the code selected, and use "Insert at cursor", it replaces the code itself with the documentation 😅

toolmantim commented:
I just ran another test if it helps. `/explain` seems to work well on a whole file, but `/doc` and `/test` don't seem to work well:

<img width="1150" alt="Screenshot 2023-08-08 at 4 51 29 pm" src="https://github.com/sourcegraph/cody/assets/153/77b86806-5447-49c3-9912-6861980264a9">

<img width="1150" alt="Screenshot 2023-08-08 at 4 50 21 pm" src="https://github.com/sourcegraph/cody/assets/153/bd56dc13-ee99-47b9-9d22-783e91905913">

In theory `/doc` could work on a whole file, just like `/explain`… generating a big summary comment you might want to insert at the top?

I'm not sure how `/test` could be changed to make sense for the whole file? 🤔

toolmantim commented:
With how the errors are displayed, we already have an in-chat error mechanism for when we executing multiple recipes at once:

<img width="438" alt="Screenshot 2023-08-08 at 5 55 03 pm" src="https://github.com/sourcegraph/cody/assets/153/05dd55e8-f572-434b-9f6b-d4cee7ad312e">

We should avoid adding yet another way of showing errors. We're still going to need to use VS Code native ones for in-editor experiences (if we keep the chat sidebar hidden).

Here's a better [in-chat error design](https://www.figma.com/file/9112BsKsJc1BpO2j8XYwLL/%F0%9F%A4%96-Cody-VS-Code-%5BMain%5D?type=design&node-id=4541-152259&mode=design&t=IbaGoksebVr7FqA8-4) for something we can use for both the above error message, a `/test` error, and any custom commands that require code to be selected:

https://github.com/sourcegraph/cody/assets/153/87ab3dcb-4a16-456d-ab15-10ca920ed400

This avoids cleaning the input box, making it an expensive mistake for the user. Instead it keeps the input box and cursor unchanged, so you can go select code and easily try again… a nice fast & snappy feedback loop, helping people to just click around and try commands without it feeling too heavyweight if they make a mistake.

abeatrix commented:
Thanks for the thorough testing TIm! Much appreciated 🙇‍♀️  

> I just ran another test if it helps. /explain seems to work well on a whole file, but /doc and /test don't seem to work well

I will clarify with Beyang but I think I misunderstood his initial feedback. I think was talking about /explain specifically and not /doc and /test, so that was my bad 😅 

> We should avoid adding yet another way of showing errors.

@toolmantim `this.transcript.addErrorAsAssistantResponse(errorMsg)` is an existing to add error to display as a response from Cody. Beyang was thinking "that's the way it would show up if you executed a command in the terminal, and this is similar to that", but from design perspective, do you think displaying this error in the banner is more appropriate in this case?

P.S. We can update the existing error banner style to what you have in the design in another PR

toolmantim commented:
Yeah, I think the error banner is more appropriate in this context… even a terminal application might do smart things to not clear your REPL input buffer on validation fail. But definitely keen to hear any strong opinions otherwise!

beyang commented:
I was thinking the most natural form for error messages would just be as a response message in the REPL. Similar to how in your terminal, if you type in an invalid command, the output will be an error message:
![image](https://github.com/sourcegraph/cody/assets/1646931/22457f24-3eec-4e79-8611-f0d21eb9327c)

See also slash-commands in Slack:
![image](https://github.com/sourcegraph/cody/assets/1646931/4b4be40f-9a14-46c7-9c30-9539a13f2584)
![image](https://github.com/sourcegraph/cody/assets/1646931/c9648a9c-07b7-449f-a17f-eda7b449ba3a)

In a REPL, I expect responses—whether success or error—to print in the main text area. So if you tried to run `/doc` without a selection of code, the response could be a message like "Select some code to document and re-run /doc".

abeatrix commented:
@beyang @toolmantim I've updated the PR so that:
- when no code is selected when running `/explain`, it will use visible content automatically
- when there are selected code when running `/explain`, use selected code
- when no code is selected for `/doc` and `/test`, display error in chat instead of the vs code system error pop up
See demo video listed under Test plan for details.

I've also updated the walkthrough to replace recipes and commands.

abeatrix commented:
updated error message style from
<img width="649" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/5588f4e0-ea4f-4f8d-8a2e-e18dde6081e1">

to 
<img width="590" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/8212312e-67e0-4188-88dd-b466b2315f42">

beyang commented:
Is there a way to make the selection appear when you click on the file link, rather than just the start of the range? (This can be left for a follow-up PR.)

Current behavior:
![image](https://github.com/sourcegraph/cody/assets/1646931/cab6ac98-552c-4529-9ca7-1f88d70a0702)

Desired:
![image](https://github.com/sourcegraph/cody/assets/1646931/dd53d718-88d6-42b3-82d2-ef3401db0fe9)

beyang commented:
/doc appears to work without a selection, but it's unclear which part of the file it's documenting (no range in the file link). Without a range, my initial assumption is that it would document the whole file, but the generated documentation only seems to refer to one function in particular:
![image](https://github.com/sourcegraph/cody/assets/1646931/65a26d59-1980-4026-a4b1-2e78ee50ed91)

beyang commented:
This can be left to a follow-up PR, but we should do some basic intent detection to map queries like "explain this code" to use the `/explain` command:
![image](https://github.com/sourcegraph/cody/assets/1646931/75eadb44-ba98-4037-a4aa-b39b6249af11)

beyang commented on agent/src/editor.ts:
```diff
@@ -79,6 +79,19 @@ export class AgentEditor implements Editor {
     }
 
+    public getActiveTextEditorSelectionOrVisibleContent(): ActiveTextEditorSelection | null {
+        const document = this.activeDocument()
```
what happens if this is invoked when `this.activeDocument` is nil?

abeatrix replied:
it falls back to use `getActiveTextEditorSelection`, which would then return `undefined` (I just copied the one for `getActiveTextEditorSelectionOrEntireFile` over 😅 ).

abeatrix replied:
removed from agent

beyang commented on lib/shared/src/chat/prompts/default-prompts.json:
```diff
@@ -10,10 +10,9 @@
     "Explain Code": {
       "slashCommand": "explain",
-      "prompt": "Explain what the selected code does in simple terms. Assume the audience is a beginner programmer who has just learned the language features and basic syntax. Focus on explaining: 1) The purpose of the code 2) What input(s) it takes 3) What output(s) it produces 4) How it achieves its purpose through the logic and algorithm. 5) Any important logic flows or data transformations ha ...
+      "prompt": "Explain what the selected code does in simple terms. Assume the audience is a beginner programmer who has just learned the language features and basic syntax. Focus on explaining: 1) The purpose of the code 2) What input(s) it takes 3) What output(s) it produces 4) How it achieves its purpose through the logic and algorithm. 5) Any important logic flows or data transformations ha ...
```
This doesn't quite seem grammatical. If I were instructing a human, I would say, "If the selected code is a named symbol like a function or class, start your explanation with the name of that symbol." But we should check what yields the best results. Are there examples you can provide where this improves response quality?

What if the selection doesn't conform to the beginning of a named scope like a function or method (i.e., what if the selection is an unnamed block of code within a function body)?

abeatrix replied:
> Are there examples you can provide where this improves response quality?

Yeee. I added this because of the change to support visible content, if we don't add the last bit, sometimes it will just explain the code without letting you know which part of the code it is explaining, so I found adding this helps. Here is an example of how the result looks like without the last added bit:

<img width="2168" alt="Screenshot 2023-08-09 at 5 32 09 PM" src="https://github.com/sourcegraph/cody/assets/68532117/1ac3d10a-ebe6-44d3-bb9b-89d05af1971c">

When we specified to start with names for the same code:

<img width="2124" alt="Screenshot 2023-08-09 at 5 33 28 PM" src="https://github.com/sourcegraph/cody/assets/68532117/36fa5544-8a29-4e81-b8c5-63e632ac7d8b">


It also doesn't affect code that doesn't has a name or non-function from what I've tested so far

<img width="2168" alt="Screenshot 2023-08-09 at 5 20 28 PM" src="https://github.com/sourcegraph/cody/assets/68532117/2ba59a50-7880-4ced-9ccb-895b6c4eec14">

Here is the answer after the last bit is added:
<img width="2124" alt="Screenshot 2023-08-09 at 5 24 33 PM" src="https://github.com/sourcegraph/cody/assets/68532117/9084f377-1cb2-4f3b-b77c-a23068abdbbb">

abeatrix commented:
> /doc appears to work without a selection, but it's unclear which part of the file it's documenting (no range in the file link). Without a range, my initial assumption is that it would document the whole file, but the generated documentation only seems to refer to one function in particular: ![image](https://user-images.githubusercontent.com/1646931/259577191-65a26d59-1980-4026-a4b1-2e78ee50ed91.png)
> 
> Ditto for /test

I've been trying to reproduce this but have yet to be successful... how did you run those commands?

abeatrix commented:
@toolmantim do you mind checking if `/doc` and `/test` commands work for you when you build from this branch? 🙇‍♀️

toolmantim commented:
Yep, I missed this yesterday sorry! Will do.

toolmantim commented:
The bit of UI under the message box is meant to represent what context/subject is being used to the user, but it doesn't seem to match anymore w/ using visible content:

https://github.com/sourcegraph/cody/assets/153/ebd5abbf-0534-4d31-9a6a-a95388b7c162

Is that intentional?

toolmantim commented:
Any reason we're not using the cursor position? (but then invisibly sending some +/- number of lines along behind the scenes?) Because then that bit of UI could just show `configuration.test.ts:99` when your cursor is on the `expect(getConfiguration(config)).toEqual({` line in that video, and the user has a basic understanding of what's going on still ("This is the line I care about / want explained… even if you have to do some magic to make that work…")

abeatrix commented:
> The bit of UI under the message box is meant to represent what context/subject is being used to the user, but it doesn't seem to match anymore w/ using visible content:

I'd say that's the expected behavior because the one in the footer tracks live range v.s. the one being used in file is the range that is being used as context.

beyang commented on vscode/CHANGELOG.md:
```diff
@@ -19,6 +20,8 @@ Starting from `0.2.0`, Cody is using `major.EVEN_NUMBER.patch` for release versi
 
 ### Changed
 
+- Update shortcut for Cody Commands Menu to `alt` + `c` due to conflict with exisiting keybind for `fixup`. [pull/602](https://github.com/sourcegraph/cody/pull/602)
```
```suggestion
- Update shortcut for Cody Commands Menu to `alt` + `c` due to conflict with existing keybinding for `fixup`. [pull/602](https://github.com/sourcegraph/cody/pull/602)
```

toolmantim commented:
> do you mind checking if /doc and /test commands work for you when you build from this branch? 🙇‍♀️

@abeatrix tested, and they both seem to work 👍🏼

abeatrix commented:
@toolmantim Wooooop sorry i meant to ask if you can check if they work when you have no selection in the doc.

toolmantim commented:
@beyang I hear you on the REPL error style. This felt different to me, as we're asking them to juggle two terminals (chat pane and editor pane) and I was trying to avoid them having to retype and retry in that juggle. But what's implemented here is fine! But could we focus on making it so the user rarely even sees the error condition at all?

After testing this some more… I think visible content support might be moving us away from making it simpler to use Cody. Or at least a distraction for some other fixes?

If you just right click somewhere in your code and run one of the Cody commands, each of them now work very differently:

- `/doc` will show an error
- `/test` will show an error
- `/smell` will use visible content
-  `/explain` will use visible content
- "Refactor Code" will use cursor position

https://github.com/sourcegraph/cody/assets/153/c09d9661-1678-4e77-ab4e-2fb5d7643212

We should move towards making them more consistent, and have the common cases just work first time vs show an error. Your first time Cody experience shouldn't require an error message, and retyping & submitting an input.

I think that would mean making sure we support code selection, cursor position, or whole file (Copilot and others support whole file w/o selection for test and explain).

We'd also then have the context UI below the message box reflect either the selection, cursor position, or whole file (esp. important for cursor position… because it stops flashing in the editor when you're focused on chat).

toolmantim commented:
@abeatrix np! Just retested, but appears not (if I tested correctly — I just made sure I had no selection and then ran `/test` and `/doc`):

<img width="1207" alt="Screenshot 2023-08-10 at 4 14 21 pm" src="https://github.com/sourcegraph/cody/assets/153/ef325213-ce82-424d-8e90-5984e5a980a1">

abeatrix commented:
@toolmantim thank you!! That's the behavior I got and expecting too, but somehow works differently on beyang's. Will continue looking into it more tomorrow. Thanks you both again for reviewing!

beyang commented on lib/shared/src/chat/prompts/utils.ts:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Gets the current directory path from the file path param
+ */
+export const getCurrentDirPath = (filePath: string): string => path.dirname(filePath)
```
remove

beyang commented on lib/shared/src/chat/prompts/utils.ts:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Gets the name of the parent directory from a directory path.
+ */
+export const getParentDirName = (dirPath: string): string => path.basename(path.dirname(dirPath))
```
remove

beyang commented on lib/shared/src/chat/prompts/utils.ts:
```diff
@@ -0,0 +1,91 @@
+    return promptText
+}
+
+export async function makeInteraction(
```
createInteraction or newInteraction

beyang commented on lib/shared/src/chat/prompts/utils.ts:
```diff
@@ -0,0 +1,91 @@
+ * @param displayText - Optional human-readable display text for the request.
+ * @returns A Promise resolving to the Interaction object.
+ */
+export async function interactionWithAssistantError(errorMsg: string, displayText = ''): Promise<Interaction> {
```
"createInteractionWithError" or "newInteractionWithError"

beyang commented on vscode/src/chat/MessageProvider.ts:
```diff
@@ -556,19 +562,51 @@ export abstract class MessageProvider extends MessageHandler implements vscode.D
+    ): Promise<void> {
+        const newInteraction =
+            interaction ||
+            new Interaction(
```
use constructor function here?

beyang commented on lib/shared/src/chat/prompts/index.ts:
```diff
@@ -66,6 +66,7 @@ export interface CodyPromptContext {
     filePath?: string
     directoryPath?: string
     none?: boolean
+    strict?: boolean // add additional prompt at the end to prevent hallucination
```
remove?

beyang commented on lib/shared/src/chat/prompts/default-prompts.json:
```diff
@@ -10,10 +10,9 @@
       "slashCommand": "explain",
-      "prompt": "Explain what the selected code does in simple terms. Assume the audience is a beginner programmer who has just learned the language features and basic syntax. Focus on explaining: 1) The purpose of the code 2) What input(s) it takes 3) What output(s) it produces 4) How it achieves its purpose through the logic and algorithm. 5) Any important logic flows or data transformations ha ...
+      "prompt": "Explain what the selected code does in simple terms. Assume the audience is a beginner programmer who has just learned the language features and basic syntax. Focus on explaining: 1) The purpose of the code 2) What input(s) it takes 3) What output(s) it produces 4) How it achieves its purpose through the logic and algorithm. 5) Any important logic flows or data transformations ha ...
       "context": {
```
`s/context/requiredContext/` perhaps?

abeatrix replied:
Keeping this as it is for now if that's not an issue, as we already have users setting up commands with this field.

beyang commented on lib/shared/src/chat/recipes/custom-prompt.ts:
```diff
@@ -118,303 +116,75 @@ export class CustomPrompt implements Recipe {
     ): Promise<ContextMessage[]> {
         const contextMessages: ContextMessage[] = []
 
+        // NONE
```
Lower case the inline comments

beyang commented on lib/shared/src/chat/recipes/custom-prompt.ts:
```diff
@@ -44,68 +50,60 @@ export class CustomPrompt implements Recipe {
-        // Prompt text to share with Cody only and not display to human
-        const codyPromptText = selectionPromptText + prompts.instruction.replace('{humanInput}', promptText)
+        const selection = selectionContent || context.editor.controllers?.inline?.selection
+        if (isContextNeeded?.selection === true && !selection?.selectedText) {
```
```suggestion
        if (isContextNeeded?.selection && !selection?.selectedText) {
```

beyang commented on lib/shared/src/chat/recipes/custom-prompt.ts:
```diff
@@ -44,68 +50,60 @@ export class CustomPrompt implements Recipe {
-                : ''
+        // Check if selection is required. If selection is not defined, accept visible content
+        const selectionContent =
+            isContextNeeded?.selection === true
```
```suggestion
            isContextNeeded?.selection
```
